File: sdkwork-rtc-sdk-go/rtcstandard/provider_support.go
package rtcstandard

type RtcProviderSupportStateRequest struct {
    ProviderKey string
    Builtin     bool
    Official    bool
    Registered  bool
}

type RtcProviderSupport struct {
    ProviderKey string
    Status      string
    Builtin     bool
    Official    bool
    Registered  bool
}

var RtcProviderSupportStatuses = []string{
    "builtin_registered",
    "official_registered",
    "official_unregistered",
    "unknown",
}

func ResolveRtcProviderSupportStatus(request RtcProviderSupportStateRequest) string {
    if request.Official && request.Registered {
        if request.Builtin {
            return "builtin_registered"
        }
        return "official_registered"
    }

    if request.Official {
        return "official_unregistered"
    }

    return "unknown"
}

func CreateRtcProviderSupportState(
    request RtcProviderSupportStateRequest,
) RtcProviderSupport {
    return RtcProviderSupport{
        ProviderKey: request.ProviderKey,
        Status:      ResolveRtcProviderSupportStatus(request),
        Builtin:     request.Builtin,
        Official:    request.Official,
        Registered:  request.Registered,
    }
}

File: sdkwork-rtc-sdk-rust/src/provider_support.rs
#[allow(non_snake_case)]
pub struct RtcProviderSupport {
    pub providerKey: String,
    pub status: &'static str,
    pub builtin: bool,
    pub official: bool,
    pub registered: bool,
}

#[allow(non_snake_case)]
pub struct RtcProviderSupportStateRequest {
    pub providerKey: String,
    pub builtin: bool,
    pub official: bool,
    pub registered: bool,
}

pub const RTC_PROVIDER_SUPPORT_STATUSES: [&str; 4] = [
    "builtin_registered",
    "official_registered",
    "official_unregistered",
    "unknown",
];

pub fn resolve_rtc_provider_support_status(
    request: &RtcProviderSupportStateRequest,
) -> &'static str {
    if request.official && request.registered {
        return if request.builtin {
            "builtin_registered"
        } else {
            "official_registered"
        };
    }

    if request.official {
        return "official_unregistered";
    }

    "unknown"
}

pub fn create_rtc_provider_support_state(
    request: RtcProviderSupportStateRequest,
) -> RtcProviderSupport {
    let status = resolve_rtc_provider_support_status(&request);

    RtcProviderSupport {
        providerKey: request.providerKey,
        status,
        builtin: request.builtin,
        official: request.official,
        registered: request.registered,
    }
}
